### PR TITLE
add harvesterVersion

### PR DESCRIPTION
**Pull request by FrankYang0529**
Add a new field `HarvesterVersion` in `CheckUpgradeRequest`. Also, add a `GetVersion` function to handle `AppVersion` and `HarvesterVersion`.

https://github.com/harvester/harvester/issues/1849

### CODE REVIEW

File: upgraderesponder/service.go [hunk 1]
```diff
@@ -77,8 +77,16 @@ type Version struct {
 }
 
 type CheckUpgradeRequest struct {
-	AppVersion string            `json:"appVersion"`
-	ExtraInfo  map[string]string `json:"extraInfo"`
+	AppVersion       string            `json:"appVersion"`
+	HarvesterVersion string            `json:"harvesterVersion"`
```
Comment (PhanLe1010) [inline upgraderesponder/service.go:81]: Can we add a comment explaining why we need this workaround and when we can remove it?

File: upgraderesponder/service.go [hunk 2]
```diff
@@ -77,8 +77,16 @@ type Version struct {
...
+	ExtraInfo        map[string]string `json:"extraInfo"`
+}
+
+func (r *CheckUpgradeRequest) GetVersion() string {
+	if r.AppVersion != "" || r.HarvesterVersion == "" {
+		return r.AppVersion
+	}
+	return r.HarvesterVersion
```
Comment (PhanLe1010) [inline upgraderesponder/service.go:89]: Is there different beetween the above and this logic?
```suggestion
	if r.AppVersion != "" {
		return r.AppVersion
	}
	return r.HarvesterVersion
```

File: upgraderesponder/service.go (hunk 1 above)
Comment (FrankYang0529) [reply upgraderesponder/service.go:84]: Added it. Thank you.

File: upgraderesponder/service.go (hunk 2 above)
Comment (FrankYang0529) [reply upgraderesponder/service.go]: I change the logic to the following. Do you think it's more clear?  Thank you.

```go
// GetVersion returns version from AppVersion and HarvesterVersion.
// Only if AppVersion is empty and HarvesterVersion is not, we return HarvesterVersion.
// We will remove this function after all customers' Harvester version update to v1.1.0.
func (r *CheckUpgradeRequest) GetVersion() string {
	if r.AppVersion == "" && r.HarvesterVersion != "" {
		return r.HarvesterVersion
	}
	return r.AppVersion
}
```

Comment (PhanLe1010) [reply upgraderesponder/service.go]: I the future, Longhorn also need to migrate to this newer format. So, how di you think about 

```go
// GetVersion returns version from AppVersion and HarvesterVersion.
// Only if AppVersion is empty and HarvesterVersion is not, we return HarvesterVersion.
// We will remove this function after all customers' Harvester version update to v1.1.0.
func (r *CheckUpgradeRequest) GetVersion() string {
	if r.AppVersion != "" {
		return r.AppVersion
	}
        // for future Longhorn migration
        //  if r.LonghornVersion != "" {
	//	 return r.LonghornVersion
	// }
        if r.HarvesterVersion != "" {
		return r.HarvesterVersion
	}
	return ""

}
```

Comment (FrankYang0529) [reply upgraderesponder/service.go]: Yes, this one looks good to me. I update it. Thank you.

**Review by PhanLe1010 — APPROVED** [review]

**Pull request merged**